src/models/pinns_ivfm.py

```diff
@@ -0,0 +1,212 @@
+from functools import partial
+
+import numpy as np
+import torch
+from torch import nn
+from tqdm.auto import tqdm
+
+from src.utils.automatic_differentiation import grad
+
+
+class PINNs_ivfm(nn.Module):
+    """Neural network with an MLP architecture."""
+
+    # Define the constructor class with inheritance
+    def __init__(
+        self,
+        mlp: nn.Module,
+        epochs: int = 1000,
+        loss: nn.Module = nn.L1Loss(),
+        optimizer: torch.optim.Optimizer = partial(torch.optim.Adam, lr=1e-3),
+        physical_loss: bool = True,
+        lambda_div: float = 1,
+        lambda_BC: float = 1,
+        lambda_smooth: float = 1,
+    ) -> None:
+        """Initialize class instance.
+
+        Args:
+            mlp: MLP network.
+            epochs: Number of epochs.
+            loss: Reconstruction loss function.
+            optimizer: Optimizer.
+            physical_loss: Whether to use physical loss.
+            lambda_div: Weight for free divergence residual loss.
+            lambda_BC: Weight for boundary condition residual loss.
+            lambda_smooth: Weight for smoothing regularization.
+        """
+        super().__init__()
+
+        # Define attributes
+        self.epochs = epochs
+        self.loss = loss
+        self.mlp = mlp
+        self.optimizer = optimizer(self.mlp.parameters())
+        self.lambda_div = lambda_div  # Weight for divergence residual
+        self.lambda_BC = lambda_BC  # Weight for boundary residual
+        self.lambda_smooth = lambda_smooth  # Weight for smoothing regularization
+        self.do_phy_loss = physical_loss
+
+    def np_to_th(self, x: np.ndarray, device: torch.device) -> torch.Tensor:
+        """Convert numpy data to torch.Tensor with the selected DEVICE (cpu or gpu).
+
+        Args:
+            x: Numpy data.
+
+        Returns:
+            Converted torch.Tensor data.
+        """
+        n_samples = len(x)
+        return torch.from_numpy(x).float().to(device).reshape(n_samples, -1)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass."""
+        return self.mlp(x)
+
+    def physical_loss(self, batch: dict[str, torch.Tensor], device) -> torch.Tensor:
+        """Compute physical loss."""
+        grid_R = self.np_to_th(batch["grid_R"], device).requires_grad_(True)
+        grid_TH = self.np_to_th(batch["grid_TH"], device).requires_grad_(True)
+        # uD = self.np_to_th(batch["uD"], device)
+
+        x = torch.hstack((grid_R, grid_TH))
+        # x = torch.hstack((grid_R, grid_TH, uD))
+
+        # Compute the corresponding estimated velocities
+        preds = self.forward(x)
```

Comment:
I just wondering the correctness of applying two forward passes consecutively, one for the data attachment loss and one here for the physical loss. This is not a comment to change the code, but more a generic question to be discussed in our upcoming meetings ;)

Reply:
This I understand, but why not using only one single forward pass ? any idea ? To be discussed ;)